numpy/lib/function_base.py

```diff
@@ -912,49 +923,84 @@ def piecewise(x, condlist, funclist, *args, **kw):
+           [ 6.25,  3.75,  1.25,  1.25,  3.75,  6.25]])
     """
-    x = asanyarray(x)
-    n2 = len(funclist)
-    if (isscalar(condlist) or not (isinstance(condlist[0], list) or
-                                   isinstance(condlist[0], ndarray))):
-        condlist = [condlist]
-    condlist = array(condlist, dtype=bool)
-    n = len(condlist)
-    # This is a hack to work around problems with NumPy's
-    #  handling of 0-d arrays and boolean indexing with
-    #  numpy.bool_ scalars
-    zerod = False
-    if x.ndim == 0:
-        x = x[None]
-        zerod = True
-        if condlist.shape[-1] != 1:
-            condlist = condlist.T
-    if n == n2 - 1:  # compute the "otherwise" condition.
-        totlist = np.logical_or.reduce(condlist, axis=0)
-        condlist = np.vstack([condlist, ~totlist])
-        n += 1
-    if (n != n2):
-        raise ValueError(
-                "function list and condition list must be the same")
+    def otherwise_condition(condlist):
+        return ~np.logical_or.reduce(condlist, axis=0)
 
-    y = zeros(x.shape, x.dtype)
-    for k in range(n):
-        item = funclist[k]
-        if not isinstance(item, collections.Callable):
-            y[condlist[k]] = item
-        else:
-            vals = x[condlist[k]]
-            if vals.size > 0:
-                y[condlist[k]] = item(vals, *args, **kw)
-    if zerod:
-        y = y.squeeze()
-    return y
+    def ok_shapes(condlist, funclist):
+        nc, nf = len(condlist), len(funclist)
+        return nc in [nf-1, nf]
+
+    def check_shapes(condlist, funclist):
+        if not ok_shapes(condlist, funclist):
+            raise ValueError("function list and condition list" +
+                             " must be the same length")
+
+    if (np.isscalar(condlist) or not ok_shapes(condlist, funclist)):
+        condlist = (condlist,)
+    check_shapes(condlist, funclist)
+    if not isinstance(xi, tuple):
+        xi = (xi,)
+
+    condlist = np.broadcast_arrays(*condlist)
+    if len(condlist) == len(funclist)-1:
+        condlist.append(otherwise_condition(condlist))
+    # If cond array is not an ndarray in boolean format or scalar bool, abort.
+    deprecated_ints = False
+    for i, cond in enumerate(condlist):
+        if cond.dtype.type is not np.bool_:
+            if np.issubdtype(cond.dtype, np.integer):
+                # A previous implementation accepted int ndarrays accidentally.
+                # Supported here deliberately, but deprecated.
+                condlist[i] = condlist[i].astype(bool)
+                deprecated_ints = True
+            else:
+                raise ValueError(
+                    'invalid entry in condlist: should be boolean ndarray')
+
+    if deprecated_ints:
+        msg = "piecewise condlists containing integer ndarrays is deprecated " \
+            "and will be removed in the future. Use `.astype(bool)` to " \
+            "convert to bools."
+        warnings.warn(msg, DeprecationWarning)
```

Comment:
This change is probably sensible to extract to its own PR. Note this is missing `stacklevel=2`.